trie/fog.py

```diff
@@ -0,0 +1,284 @@
+                # Raising a IndexError is the desired behavior in this case!
+                return self._unexplored_prefixes[index]
+
+    @staticmethod
+    @to_tuple
+    def _prefix_distance(low_key: Nibbles, high_key: Nibbles) -> Tuple[int, ...]:
+        """
+        How far are the two keys from each other, as a sequence of differences.
+        The first non-zero distance must be positive, but the remaining distances may
+        be negative. Distances are designed to be simply compared, like distance1 < distance2.
+
+        The high_key must be higher than the low key, or the output distances are not
+        guaranteed to be accurate.
+        """
+        for low_nibble, high_nibble in zip_longest(low_key, high_key, fillvalue=None):
+            if low_nibble is None:
+                final_low_nibble = 15
+            else:
+                final_low_nibble = low_nibble
+
+            if high_nibble is None:
+                final_high_nibble = 0
+            else:
+                final_high_nibble = high_nibble
+
+            # Note: this might return a negative value. It's fine, because only the
+            #   relative distance matters. For example (1, 2) and (2, 1) produce a
+            #   distance of (1, -1). If the other reference point is (3, 1), making
+            #   the distance to the middle (1, 0), then the "correct" thing happened.
+            #   The (1, 2) key is a tiny bit closer to the (2, 1) key, and a tuple
+            #   comparison of the distance will show it as a smaller distance.
+            yield final_high_nibble - final_low_nibble
+
+    @classmethod
+    def _new_trie_fog(cls, unexplored_prefixes: SortedSet) -> 'HexaryTrieFog':
+        """
+        Convert a set of unexplored prefixes to a proper HexaryTrieFog object.
+        """
+        copy = cls()
+        copy._unexplored_prefixes = unexplored_prefixes
+        return copy
+
+    def serialize(self) -> bytes:
+        prefixes = list(self._unexplored_prefixes)
+        return f"HexaryTrieFog:{prefixes!r}".encode()
+
+    @classmethod
+    def deserialize(cls, encoded: bytes) -> 'HexaryTrieFog':
+        serial_prefix = b'HexaryTrieFog:'
+        if not encoded.startswith(serial_prefix):
+            raise ValueError(f"Cannot deserialize this into HexaryTrieFog object: {encoded}")
+        else:
+            encoded_list = encoded[len(serial_prefix):]
+            prefix_list = ast.literal_eval(encoded_list.decode())
+            # confirm that all prefixes are valid nibbles, while building the sorted set
+            valid_prefixes = (Nibbles(prefix) for prefix in prefix_list)
+            # build the Trie Fog instance
+            deserialized_prefixes = SortedSet(valid_prefixes)
+            return cls._new_trie_fog(deserialized_prefixes)
+
+    def __eq__(self, other: Any) -> bool:
+        if not isinstance(other, HexaryTrieFog):
+            return False
```

Comment:
I would say this should raise a `TypeError` because we are probably doing something wrong if we are comparing to the wrong class.

Reply:
I would be pretty surprised to get a `TypeError` at the console from `1 == 'a'`. I think the surprise of `==` raising an exception is probably worse than allowing a user to accidentally equality test against the wrong thing. In a quick search, I couldn't find anyone online that recommended raising a `TypeError` from `__eq__`.

Reply:
I think what I may have been remembering that inspired me to do this was a SQLAlchemy API, but now that I'm remembering it was in the query API where you construct the query constraints like `Session.query(User).filter(User.name == 'bob')` where the `User.name == 'bob'` actually returns a non-boolean `Constraint` object which is later translated into the SQL query.  In this case it makes sense to raise exceptions when you do something like `User.name == 1234` since `1234` is not a string and thus the API doesn't make sense when you pass an object of the wrong type....

Long story short, :+1: